File: apple/ClueweaveKit/Sources/ClueweaveKit/Solver.swift
// Whole-grid solver: propagation to fixpoint + bounded search
// (port of src/engine/solver.ts).

public enum SolveStatus: Sendable {
    case solved, stuck, contradiction
}

public func makeGrid(_ h: Int, _ w: Int) -> Grid {
    Array(repeating: Array(repeating: Cell.unknown, count: w), count: h)
}

func column(of grid: Grid, _ c: Int) -> [Cell] {
    grid.map { $0[c] }
}

func isComplete(_ grid: Grid) -> Bool {
    grid.allSatisfy { row in row.allSatisfy { $0 != .unknown } }
}

/// Iterated line solving across all rows and columns to a fixpoint. Pure logical
/// propagation — never guesses.
public func propagate(
    _ rowClues: [Clue], _ colClues: [Clue], start: Grid? = nil
) -> (status: SolveStatus, grid: Grid) {
    let h = rowClues.count
    let w = colClues.count
    var grid = start ?? makeGrid(h, w)

    var changed = true
    while changed {
        changed = false
        for r in 0..<h {
            guard let next = solveLine(grid[r], rowClues[r]) else {
                return (.contradiction, grid)
            }
            for c in 0..<w where next[c] != grid[r][c] {
                grid[r][c] = next[c]
                changed = true
            }
        }
        for c in 0..<w {
            guard let next = solveLine(column(of: grid, c), colClues[c]) else {
                return (.contradiction, grid)
            }
            for r in 0..<h where next[r] != grid[r][c] {
                grid[r][c] = next[r]
                changed = true
            }
        }
    }

    return (isComplete(grid) ? .solved : .stuck, grid)
}

func firstUnknown(_ grid: Grid) -> (Int, Int)? {
    for r in 0..<grid.count {
        for c in 0..<grid[r].count where grid[r][c] == .unknown {
            return (r, c)
        }
    }
    return nil
}

/// Safety cap so a pathological grid can never hang the UI.
let nodeCap = 200_000

public struct SolutionCount: Sendable {
    public let count: Int
    /// True if the node cap was hit — the count is a lower bound and
    /// uniqueness must NOT be inferred.
    public let capped: Bool
}

/// Count solutions up to `limit` via propagation interleaved with DFS on the
/// first unknown cell. Node-bounded.
public func countSolutionsDetailed(_ rowClues: [Clue], _ colClues: [Clue], limit: Int = 2) -> SolutionCount {
    var count = 0
    var nodes = 0
    var capped = false

    func dfs(_ grid: Grid) {
        if count >= limit { return }
        nodes += 1
        if nodes > nodeCap {
            capped = true
            return
        }
        let res = propagate(rowClues, colClues, start: grid)
        if res.status == .contradiction { return }
        if res.status == .solved {
            count += 1
            return
        }
        guard let (r, c) = firstUnknown(res.grid) else { return }
        for value in [Cell.filled, Cell.empty] {
            if count >= limit { return }
            var next = res.grid
            next[r][c] = value
            dfs(next)
        }
    }

    dfs(makeGrid(rowClues.count, colClues.count))
    return SolutionCount(count: count, capped: capped)
}

public func hasUniqueSolution(_ rowClues: [Clue], _ colClues: [Clue]) -> Bool {
    let result = countSolutionsDetailed(rowClues, colClues, limit: 2)
    return result.count == 1 && !result.capped
}

/// Return up to `limit` actual solution grids (booleans). Node-bounded.
public func enumerateSolutions(_ rowClues: [Clue], _ colClues: [Clue], limit: Int = 2) -> [[[Bool]]] {
    var out: [[[Bool]]] = []
    var nodes = 0
    func dfs(_ grid: Grid) {
        if out.count >= limit { return }
        nodes += 1
        if nodes > nodeCap { return }
        let res = propagate(rowClues, colClues, start: grid)
        if res.status == .contradiction { return }
        if res.status == .solved {
            out.append(toBooleans(res.grid))
            return
        }
        guard let (r, c) = firstUnknown(res.grid) else { return }
        for value in [Cell.filled, Cell.empty] {
            if out.count >= limit { return }
            var next = res.grid
            next[r][c] = value
            dfs(next)
        }
    }
    dfs(makeGrid(rowClues.count, colClues.count))
    return out
}

/// True if pure propagation (no guessing) fully solves the puzzle.
public func isLineSolvable(_ rowClues: [Clue], _ colClues: [Clue]) -> Bool {
    propagate(rowClues, colClues).status == .solved
}

func toBooleans(_ grid: Grid) -> [[Bool]] {
    grid.map { row in row.map { $0 == .filled } }
}

File: web/src/engine/solver.ts
import { UNKNOWN, FILLED, EMPTY, type Cell, type Clue } from "./types";
import { solveLine } from "./lineSolver";

export type Grid = Cell[][];
export type SolveStatus = "solved" | "stuck" | "contradiction";

export function makeGrid(h: number, w: number): Grid {
  return Array.from({ length: h }, () => Array<Cell>(w).fill(UNKNOWN));
}

function getColumn(grid: Grid, c: number): Cell[] {
  return grid.map((row) => row[c]);
}

function isComplete(grid: Grid): boolean {
  return grid.every((row) => row.every((cell) => cell !== UNKNOWN));
}

/**
 * Iterated line solving across all rows and columns to a fixpoint. Pure logical
 * propagation — never guesses. Accepts an optional starting grid.
 */
export function propagate(
  rowClues: Clue[],
  colClues: Clue[],
  start?: Grid,
): { status: SolveStatus; grid: Grid } {
  const h = rowClues.length;
  const w = colClues.length;
  const grid = start ? start.map((row) => row.slice()) : makeGrid(h, w);

  let changed = true;
  while (changed) {
    changed = false;
    for (let r = 0; r < h; r++) {
      const next = solveLine(grid[r], rowClues[r]);
      if (next === null) return { status: "contradiction", grid };
      for (let c = 0; c < w; c++) {
        if (next[c] !== grid[r][c]) {
          grid[r][c] = next[c];
          changed = true;
        }
      }
    }
    for (let c = 0; c < w; c++) {
      const next = solveLine(getColumn(grid, c), colClues[c]);
      if (next === null) return { status: "contradiction", grid };
      for (let r = 0; r < h; r++) {
        if (next[r] !== grid[r][c]) {
          grid[r][c] = next[r];
          changed = true;
        }
      }
    }
  }

  return { status: isComplete(grid) ? "solved" : "stuck", grid };
}

function firstUnknown(grid: Grid): [number, number] | null {
  for (let r = 0; r < grid.length; r++) {
    for (let c = 0; c < grid[r].length; c++) {
      if (grid[r][c] === UNKNOWN) return [r, c];
    }
  }
  return null;
}

/** Safety cap so a pathological grid can never hang the UI. */
const NODE_CAP = 200_000;

export interface SolutionCount {
  count: number;
  /** True if the node cap was hit before the search finished — the count is a
   *  lower bound and uniqueness must NOT be inferred. */
  capped: boolean;
}

/**
 * Count solutions up to `limit` via propagation interleaved with DFS on the
 * first unknown cell. Node-bounded so a pathological grid can't hang; if the
 * cap is hit, `capped` is set so callers never certify uniqueness they didn't
 * actually prove.
 */
export function countSolutionsDetailed(rowClues: Clue[], colClues: Clue[], limit = 2): SolutionCount {
  let count = 0;
  let nodes = 0;
  let capped = false;

  function dfs(grid: Grid): void {
    if (count >= limit) return;
    if (++nodes > NODE_CAP) {
      capped = true;
      return;
    }
    const res = propagate(rowClues, colClues, grid);
    if (res.status === "contradiction") return;
    if (res.status === "solved") {
      count++;
      return;
    }
    const cell = firstUnknown(res.grid);
    if (!cell) return;
    const [r, c] = cell;
    for (const value of [FILLED, EMPTY] as Cell[]) {
      if (count >= limit) return;
      const next = res.grid.map((row) => row.slice());
      next[r][c] = value;
      dfs(next);
    }
  }

  dfs(makeGrid(rowClues.length, colClues.length));
  return { count, capped };
}

export function countSolutions(rowClues: Clue[], colClues: Clue[], limit = 2): number {
  return countSolutionsDetailed(rowClues, colClues, limit).count;
}

export function hasUniqueSolution(rowClues: Clue[], colClues: Clue[]): boolean {
  const { count, capped } = countSolutionsDetailed(rowClues, colClues, 2);
  return count === 1 && !capped;
}

/** Return up to `limit` actual solution grids (booleans). Node-bounded. */
export function enumerateSolutions(rowClues: Clue[], colClues: Clue[], limit = 2): boolean[][][] {
  const out: boolean[][][] = [];
  let nodes = 0;
  function dfs(grid: Grid): void {
    if (out.length >= limit) return;
    if (++nodes > NODE_CAP) return;
    const res = propagate(rowClues, colClues, grid);
    if (res.status === "contradiction") return;
    if (res.status === "solved") {
      out.push(toBooleans(res.grid));
      return;
    }
    const cell = firstUnknown(res.grid);
    if (!cell) return;
    const [r, c] = cell;
    for (const value of [FILLED, EMPTY] as Cell[]) {
      if (out.length >= limit) return;
      const next = res.grid.map((row) => row.slice());
      next[r][c] = value;
      dfs(next);
    }
  }
  dfs(makeGrid(rowClues.length, colClues.length));
  return out;
}

/**
 * Return a solution grid and whether search beyond pure propagation was needed.
 * Throws if the clues admit no solution at all.
 */
export function solve(
  rowClues: Clue[],
  colClues: Clue[],
): { solution: boolean[][]; viaSearch: boolean } {
  const prop = propagate(rowClues, colClues);
  if (prop.status === "solved") {
    return { solution: toBooleans(prop.grid), viaSearch: false };
  }

  let found: Grid | null = null;
  function dfs(grid: Grid): boolean {
    const res = propagate(rowClues, colClues, grid);
    if (res.status === "contradiction") return false;
    if (res.status === "solved") {
      found = res.grid;
      return true;
    }
    const cell = firstUnknown(res.grid);
    if (!cell) return false;
    const [r, c] = cell;
    for (const value of [FILLED, EMPTY] as Cell[]) {
      const next = res.grid.map((row) => row.slice());
      next[r][c] = value;
      if (dfs(next)) return true;
    }
    return false;
  }

  dfs(makeGrid(rowClues.length, colClues.length));
  if (!found) throw new Error("unsolvable puzzle");
  return { solution: toBooleans(found), viaSearch: true };
}

/** True if pure propagation (no guessing) fully solves the puzzle. */
export function isLineSolvable(rowClues: Clue[], colClues: Clue[]): boolean {
  return propagate(rowClues, colClues).status === "solved";
}

function toBooleans(grid: Grid): boolean[][] {
  return grid.map((row) => row.map((cell) => cell === FILLED));
}
